### PR TITLE
Represent native contracts as interop packages

### PR DESCRIPTION
**Pull request by fyrchik**
Implement first part of #1691 (without inlining).
Also allow more type conversions to take place.

### CODE REVIEW

File: pkg/interop/native/gas/gas.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+// TotalSupply represents `totalSupply` method of GAS native contract.
+func TotalSupply() int {
+	return contract.Call(interop.Hash160(Hash), "totalSupply", contract.ReadStates).(int)
+}
+
+// BalanceOf represents `balanceOf` method of GAS native contract.
+func BalanceOf(addr interop.Hash160) int {
+	return contract.Call(interop.Hash160(Hash), "", contract.ReadStates, addr).(int)
```
Comment (AnnaShaleva) [inline pkg/interop/native/gas/gas.go:28]: Method name is missing.

File: pkg/interop/native/gas/gas.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+// Transfer represents `transfer` method of GAS native contract.
+func Transfer(from, to interop.Hash160, amount int) bool {
+	return contract.Call(interop.Hash160(Hash), "transfer", contract.All, from, to, amount).(bool)
```
Comment (AnnaShaleva) [inline pkg/interop/native/gas/gas.go:33]: `data` parameter is missing.

File: pkg/interop/native/neo/neo.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+// TotalSupply represents `totalSupply` method of NEO native contract.
+func TotalSupply() int {
+	return contract.Call(interop.Hash160(Hash), "totalSupply", contract.ReadStates).(int)
+}
+
+// BalanceOf represents `balanceOf` method of NEO native contract.
+func BalanceOf(addr interop.Hash160) int {
+	return contract.Call(interop.Hash160(Hash), "", contract.ReadStates, addr).(int)
```
Comment (AnnaShaleva) [inline pkg/interop/native/neo/neo.go:28]: Method name is missing.

File: pkg/interop/native/neo/neo.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+// Transfer represents `transfer` method of NEO native contract.
+func Transfer(from, to interop.Hash160, amount int) bool {
+	return contract.Call(interop.Hash160(Hash), "transfer", contract.All, from, to, amount).(bool)
```
Comment (AnnaShaleva) [inline pkg/interop/native/neo/neo.go:33]: `data` parameter is missing.

File: pkg/interop/native/neo/neo.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+// GetCommittee represents `getCommittee` method of NEO native contract.
+func GetCommittee() []interop.PublicKey {
```
Comment (AnnaShaleva) [inline pkg/interop/native/neo/neo.go:37]: Can we also include `getCandidates`, `getNextBlockValidators` and `getGasPerBlock`?

**Review by AnnaShaleva — COMMENTED** [review]
Could you also add Notary contract interop package? `balanceOf`,  `expirationOf` and `getMaxNotValidBeforeDelta` are suit.

File: pkg/compiler/codegen.go [hunk 6]
```diff
@@ -1242,6 +1249,24 @@ func (c *codegen) processDefers() {
...
 }
 
+// emitExplicitConvert handles `someType(someValue)` conversions between string/[]byte.
+// Rules for conversion:
+// 1. interop.* types are converted to ByteArray if not already.
+// 2. Otherwise convert between ByteArray/Buffer.
+// 3. Rules for types which are not string/[]byte should already
+//    be enforced by go parser.
+func (c *codegen) emitExplicitConvert(from, to types.Type) {
+	if isInteropPath(to.String()) {
```
Comment (roman-khimov) [inline pkg/compiler/codegen.go:1259]: Don't we also have structures there like `Block` or `Transaction`?

File: pkg/compiler/native_test.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+	"github.com/nspcc-dev/neo-go/pkg/interop/native/gas"
+	"github.com/nspcc-dev/neo-go/pkg/interop/native/neo"
+	"github.com/nspcc-dev/neo-go/pkg/interop/native/oracle"
+	"github.com/nspcc-dev/neo-go/pkg/interop/native/policy"
+	"github.com/stretchr/testify/require"
+)
+
+func TestContractHashes(t *testing.T) {
+	cs := native.NewContracts(false)
+	require.Equal(t, []byte(neo.Hash), cs.NEO.Hash.BytesBE())
```
Comment (roman-khimov) [inline pkg/compiler/native_test.go:17]: This allows to remove compatibility test from `native`.

File: pkg/interop/native/designate/designate.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package designate
```
Comment (roman-khimov) [inline pkg/interop/native/designate/designate.go:1]: Better use `roles`, "designate" is kinda obsolete.

File: pkg/interop/native/designate/designate.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+// Role represents node role.
+type Role byte
+
+// Various node roles.
+const (
+	StateValidator Role = 4
+	Oracle         Role = 8
+	P2PNotary      Role = 128
+)
+
```
Comment (roman-khimov) [inline pkg/interop/native/designate/designate.go:20]: `designateAsRole` support?

File: pkg/interop/native/gas/gas.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/nspcc-dev/neo-go/pkg/interop"
+	"github.com/nspcc-dev/neo-go/pkg/interop/contract"
+)
+
+// Hash represents GAS contract hash.
+const Hash = "\x3d\x2f\x0b\xce\x7a\x4d\x9b\xb5\x0b\x95\x42\x5a\x71\xf4\x44\xeb\x30\x45\x84\x75"
+
+// Symbol represents `symbol` method of GAS native contract.
+func Symbol() string {
+	return contract.Call(nil, "symbol", contract.ReadStates).(string)
```
Comment (roman-khimov) [inline pkg/interop/native/gas/gas.go:13]: Call `nil`?

File: pkg/interop/native/gas/gas.go (hunk 1 above)
Comment (roman-khimov) [inline pkg/interop/native/gas/gas.go:28]: Call `""` method?

File: pkg/interop/native/neo/neo.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/nspcc-dev/neo-go/pkg/interop"
+	"github.com/nspcc-dev/neo-go/pkg/interop/contract"
+)
+
+// Hash represents NEO contract hash.
+const Hash = "\x32\xe9\x21\xac\xc4\x75\x06\xb8\xc3\xa7\xdd\xde\xbd\x1a\x9d\x68\xca\xba\x17\xf6"
+
+// Symbol represents `symbol` method of NEO native contract.
+func Symbol() string {
+	return contract.Call(nil, "symbol", contract.ReadStates).(string)
```
Comment (roman-khimov) [inline pkg/interop/native/neo/neo.go:13]: `nil`

File: pkg/interop/native/neo/neo.go [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+// Transfer represents `transfer` method of NEO native contract.
+func Transfer(from, to interop.Hash160, amount int) bool {
+	return contract.Call(interop.Hash160(Hash), "transfer", contract.All, from, to, amount).(bool)
+}
+
+// GetCommittee represents `getCommittee` method of NEO native contract.
+func GetCommittee() []interop.PublicKey {
+	return contract.Call(interop.Hash160(Hash), "getCommittee", contract.ReadStates).([]interop.PublicKey)
+}
```
Comment (roman-khimov) [inline pkg/interop/native/neo/neo.go:39]: It has _a lot_ more public methods.

File: pkg/interop/native/policy/policy.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+// GetStoragePrice represents `getStoragePrice` method of Policy native contract.
+func GetStoragePrice() int {
+	return contract.Call(interop.Hash160(Hash), "getStoragePrice", contract.ReadStates).(int)
+}
+
+// IsBlocked represents `isBlocked` method of Policy native contract.
+func IsBlocked(addr interop.Hash160) bool {
+	return contract.Call(interop.Hash160(Hash), "isBlocked", contract.ReadStates).(bool)
+}
```
Comment (roman-khimov) [inline pkg/interop/native/policy/policy.go:44]: Setters?

Comment (roman-khimov) [reply pkg/interop/native/policy/policy.go:44]: And block/unblock account methods.

**Review by roman-khimov — CHANGES_REQUESTED** [review]
NNS and Ledger are missing completely. And this obviously is not tested.

File: pkg/interop/native/gas/gas.go (hunk 1 above)
Comment (fyrchik) [reply pkg/interop/native/gas/gas.go:28]: Fixed.

File: pkg/interop/native/gas/gas.go (hunk 2 above)
Comment (fyrchik) [reply pkg/interop/native/gas/gas.go:33]: Fixed.

File: pkg/interop/native/neo/neo.go (hunk 3 above)
Comment (fyrchik) [reply pkg/interop/native/neo/neo.go:28]: Fixed.

File: pkg/interop/native/neo/neo.go (hunk 4 above)
Comment (fyrchik) [reply pkg/interop/native/neo/neo.go:33]: Fixed.

File: pkg/interop/native/designate/designate.go (hunk 8 above)
Comment (fyrchik) [reply pkg/interop/native/designate/designate.go:1]: Fixed.

File: pkg/interop/native/designate/designate.go (hunk 9 above)
Comment (fyrchik) [reply pkg/interop/native/designate/designate.go:20]: Fixed.

File: pkg/interop/native/gas/gas.go (hunk 10 above)
Comment (fyrchik) [reply pkg/interop/native/gas/gas.go:13]: Fixed.

File: pkg/interop/native/gas/gas.go (hunk 1 above)
Comment (fyrchik) [reply pkg/interop/native/gas/gas.go:28]: Fixed.

File: pkg/interop/native/neo/neo.go (hunk 11 above)
Comment (fyrchik) [reply pkg/interop/native/neo/neo.go:13]: Fixed.

File: pkg/interop/native/policy/policy.go (hunk 13 above)
Comment (fyrchik) [reply pkg/interop/native/policy/policy.go:44]: Fixed.

File: pkg/interop/native/neo/neo.go (hunk 12 above)
Comment (fyrchik) [reply pkg/interop/native/neo/neo.go:39]: Do we need voting related methods here?

File: pkg/interop/native/neo/neo.go (hunk 11 above)
Comment (fyrchik) [reply pkg/interop/native/neo/neo.go:13]: Fixed.

File: pkg/interop/native/gas/gas.go (hunk 1 above)
Comment (fyrchik) [reply pkg/interop/native/gas/gas.go:28]: Fixed.

File: pkg/interop/native/gas/gas.go (hunk 10 above)
Comment (fyrchik) [reply pkg/interop/native/gas/gas.go:13]: Fixed.

File: pkg/interop/native/designate/designate.go (hunk 9 above)
Comment (fyrchik) [reply pkg/interop/native/designate/designate.go:20]: Fixed.

File: pkg/interop/native/designate/designate.go (hunk 8 above)
Comment (fyrchik) [reply pkg/interop/native/designate/designate.go:1]: Fixed.

File: pkg/compiler/codegen.go (hunk 6 above)
Comment (fyrchik) [reply pkg/compiler/codegen.go:1259]: Fixed.

File: pkg/compiler/native_test.go (hunk 7 above)
Comment (fyrchik) [reply pkg/compiler/native_test.go:30]: I think it is a bit different, but fixed.

File: pkg/interop/native/neo/neo.go (hunk 5 above)
Comment (fyrchik) [reply pkg/interop/native/neo/neo.go:38]: Fixed.

File: pkg/interop/native/neo/neo.go (hunk 12 above)
Comment (roman-khimov) [reply pkg/interop/native/neo/neo.go:40]: Yes.

**Review by roman-khimov — COMMENTED** [review]
It seems to be good functionally (except some methods are still missing, we need _everything_ there, take NEO voting for example, @alexvanin will be the one to ask about these missing methods), but commits can be restructured a bit, it's hard to follow them.

File: pkg/compiler/analysis.go [hunk 14]
```diff
@@ -284,3 +284,12 @@ func isInteropPath(s string) bool {
 func isNativeHelpersPath(s string) bool {
 	return strings.HasPrefix(s, interopPrefix+"/native")
 }
+
+// canConvert returns true if type doesn't need to be converted on type assertion.
+func canConvert(s string) bool {
+	if isInteropPath(s) {
+		s = s[len(interopPrefix):]
+		return s != "/iterator.Iterator"
```
Comment (AnnaShaleva) [inline pkg/compiler/analysis.go]: We also have a `storage.Context` which shouldn't be converted.

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]

File: pkg/interop/native/gas/gas.go [hunk 15]
```diff
@@ -10,12 +10,12 @@ const Hash = "\x28\xb3\xad\xab\x72\x69\xf9\xc2\x18\x1d\xb3\xcb\x74\x1e\xbf\x55\x
 
 // Symbol represents `symbol` method of GAS native contract.
 func Symbol() string {
-	return contract.Call(interop.Hash160(Hash), "symbol", contract.ReadStates).(string)
+	return contract.Call(interop.Hash160(Hash), "symbol", contract.NoneFlag).(string)
```
Comment (AnnaShaleva) [inline pkg/interop/native/gas/gas.go:13]: Should be a part of 621d5985282da510866b8f970269888c71cd7c44?

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]

File: pkg/interop/native/gas/gas.go (hunk 15 above)
Comment (fyrchik) [reply pkg/interop/native/gas/gas.go:13]: Fixed.

File: pkg/interop/native/ledger/ledger.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+	return contract.Call(interop.Hash160(Hash), "currentIndex", contract.ReadStates).(int)
+}
+
+// GetBlock represents `getBlock` method of Ledger native contract.
+func GetBlock(indexOrHash interface{}) *Block {
+	return contract.Call(interop.Hash160(Hash), "getBlock", contract.ReadStates, indexOrHash).(*Block)
+}
+
+// GetTransaction represents `getTransaction` method of Ledger native contract.
+func GetTransaction(hash interop.Hash256) *runtime.Transaction {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/interop/native/ledger/ledger.go:54]: Fixed.

File: pkg/compiler/analysis.go (hunk 14 above)
Comment (fyrchik) [reply pkg/compiler/analysis.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]
